dogstatsd-client/src/lib.rs

```diff
@@ -133,31 +129,53 @@ impl Client {
-            if let Err(err) = match action {
-                DogStatsDAction::Count(metric, value, tags) => {
-                    let metric_builder = client.count_with_tags(metric.as_ref(), value);
-                    do_send(metric_builder, tags)
-                }
-                DogStatsDAction::Distribution(metric, value, tags) => {
-                    do_send(client.distribution_with_tags(metric.as_ref(), value), tags)
-                }
-                DogStatsDAction::Gauge(metric, value, tags) => {
-                    do_send(client.gauge_with_tags(metric.as_ref(), value), tags)
-                }
-                DogStatsDAction::Histogram(metric, value, tags) => {
-                    do_send(client.histogram_with_tags(metric.as_ref(), value), tags)
-                }
-                DogStatsDAction::Set(metric, value, tags) => {
-                    do_send(client.set_with_tags(metric.as_ref(), value), tags)
+        let client_guard = match self.get_or_init_client() {
+            Ok(guard) => guard,
+            Err(e) => {
+                error!("Failed to get client: {}", e);
+                return;
+            }
+        };
+        if let Some(client) = &*client_guard {
+            for action in actions {
+                if let Err(err) = match action {
+                    DogStatsDAction::Count(metric, value, tags) => {
+                        let metric_builder = client.count_with_tags(metric.as_ref(), value);
+                        do_send(metric_builder, tags)
+                    }
+                    DogStatsDAction::Distribution(metric, value, tags) => {
+                        do_send(client.distribution_with_tags(metric.as_ref(), value), tags)
+                    }
+                    DogStatsDAction::Gauge(metric, value, tags) => {
+                        do_send(client.gauge_with_tags(metric.as_ref(), value), tags)
+                    }
+                    DogStatsDAction::Histogram(metric, value, tags) => {
+                        do_send(client.histogram_with_tags(metric.as_ref(), value), tags)
+                    }
+                    DogStatsDAction::Set(metric, value, tags) => {
+                        do_send(client.set_with_tags(metric.as_ref(), value), tags)
+                    }
+                } {
+                    error!("Error while sending metric: {}", err);
                 }
-            } {
-                error!("Error while sending metric: {}", err);
             }
         }
     }
+
+    fn get_or_init_client(&self) -> anyhow::Result<std::sync::MutexGuard<Option<StatsdClient>>> {
+        let mut client_guard = self
+            .client
+            .lock()
+            .map_err(|e| anyhow!("Failed to acquire dogstatsd client lock: {}", e))?;
+
+        if client_guard.is_none() {
+            if let Some(endpoint) = &self.endpoint {
+                *client_guard = Some(create_client(endpoint)?);
+            }
+        }
+
+        Ok(client_guard)
+    }
```

Comment:
I think that only protecting the initialization would be best. Something like this.

```suggestion
    fn get_or_init_client<'a>(&self) -> anyhow::Result<Option<Arc<StatsdClient>>> {
        if let Some(endpoint) = &self.endpoint {
            let mut guard = self.client.lock().map_err(|e| anyhow!(e.to_string()))?;
            if let Some(client) = guard.as_ref() {
                return Ok(Some(client.clone()));
            } else {
                let client = Arc::new(create_client(endpoint)?);
                *guard = Some(client.clone());
                return Ok(Some(client));
            }
        }

        Ok(None)
    }
```